crypto/src/rlwe/method.rs

```diff
@@ -0,0 +1,359 @@
+use crate::{
+    distribution::NoiseDistribution,
+    lwe::LweCiphertextMutView,
+    rlwe::structure::{
+        RlweAutoKeyMutView, RlweAutoKeyOwned, RlweAutoKeyView, RlweCiphertext,
+        RlweCiphertextMutView, RlweCiphertextView, RlweKeySwitchKey, RlweKeySwitchKeyMutView,
+        RlweKeySwitchKeyOwned, RlweKeySwitchKeyView, RlwePlaintextMutView, RlwePlaintextView,
+        RlwePublicKeyMutView, RlwePublicKeyView, RlweSecretKeyView,
+    },
+};
+use core::{borrow::Borrow, ops::Neg};
+use phantom_zone_math::{
+    decomposer::{Decomposer, DecompositionParam},
+    distribution::Ternary,
+    izip_eq,
+    misc::scratch::Scratch,
+    ring::{ElemFrom, RingOps},
+};
+use rand::RngCore;
+
+pub fn pk_gen<'a, 'b, R, T>(
+    ring: &R,
+    pk: impl Into<RlwePublicKeyMutView<'a, R::Elem>>,
+    sk: impl Into<RlweSecretKeyView<'b, T>>,
+    noise_distribution: NoiseDistribution,
+    scratch: Scratch,
+    rng: impl RngCore,
+) where
+    R: RingOps + ElemFrom<T>,
+    T: 'b + Copy,
+{
+    sk_encrypt_with_pt_in_b(
+        ring,
+        pk.into().as_ct_mut(),
+        sk,
+        noise_distribution,
+        scratch,
+        rng,
+    );
+}
+
+pub fn sk_encrypt<'a, 'b, 'c, R, T>(
+    ring: &R,
+    ct: impl Into<RlweCiphertextMutView<'a, R::Elem>>,
+    sk: impl Into<RlweSecretKeyView<'b, T>>,
+    pt: impl Into<RlwePlaintextView<'c, R::Elem>>,
+    noise_distribution: NoiseDistribution,
+    scratch: Scratch,
+    rng: impl RngCore,
+) where
+    R: RingOps + ElemFrom<T>,
+    T: 'b + Copy,
+{
+    let mut ct = ct.into();
+    ct.b_mut().copy_from_slice(pt.into().as_ref());
+    sk_encrypt_with_pt_in_b(ring, ct, sk, noise_distribution, scratch, rng);
+}
+
+pub fn sk_encrypt_with_pt_in_b<'a, 'b, R, T>(
+    ring: &R,
+    ct: impl Into<RlweCiphertextMutView<'a, R::Elem>>,
+    sk: impl Into<RlweSecretKeyView<'b, T>>,
+    noise_distribution: NoiseDistribution,
+    scratch: Scratch,
+    mut rng: impl RngCore,
+) where
+    R: RingOps + ElemFrom<T>,
+    T: 'b + Copy,
+{
+    let mut ct = ct.into();
+    let (a, b) = ct.a_b_mut();
+    ring.sample_uniform_into(a, &mut rng);
+    ring.poly_fma_elem_from(b, a, sk.into().as_ref(), scratch);
+    ring.slice_add_assign_iter(b, ring.sample_iter::<i64>(noise_distribution, &mut rng));
+}
+
+pub fn pk_encrypt<'a, 'b, 'c, R: RingOps>(
+    ring: &R,
+    ct: impl Into<RlweCiphertextMutView<'a, R::Elem>>,
+    pk: impl Into<RlwePublicKeyView<'b, R::Elem>>,
+    pt: impl Into<RlwePlaintextView<'c, R::Elem>>,
+    mut scratch: Scratch,
+    mut rng: impl RngCore,
+) {
+    let (mut ct, pk) = (ct.into(), pk.into());
+    let t0 = ring.take_eval(&mut scratch);
+    let u = ring.take_poly(&mut scratch.reborrow());
+    ring.sample_into::<i64>(u, Ternary(ring.ring_size() / 2), &mut rng);
```

Comment:
Don't assume ephemeral secret to be Ternary secret with hamming with N/2. Instead take it as a function parameter.

The reason being ephemeral secret affects noise of fresh ciphertext especially in interactive multi-party setting, hence affects noise in RGSW ciphertexts encrypted under ideal secret key used in bootstrapping key. 

Ternary distribution can be replaced with higher variance distribution in certain settings, shaving a few values off LWE dimension `n` and can be boost bootstrapping run-time.

I believe adding a new parameter will be easy (will require taking `SecertKeyDistribution` as additional parameter). But do lmk if you happen to think otherwise. We can delay the change for later.